src/kafka/connection/connection_manager.hh

```diff
@@ -0,0 +1,115 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Copyright (C) 2019 ScyllaDB Ltd.
+ */
+
+#pragma once
+
+#include "kafka_connection.hh"
+#include "../protocol/metadata_response.hh"
+#include "../protocol/metadata_request.hh"
+
+#include <boost/functional/hash.hpp>
+
+#include <unordered_map>
+
+namespace seastar {
+
+namespace kafka {
+
+struct metadata_refresh_exception : public std::runtime_error {
+public:
+    metadata_refresh_exception(const std::string& message) : runtime_error(message) {}
+};
+
+class connection_manager {
+public:
+
+    using connection_id = std::pair<std::string, uint16_t>;
+
+private:
+
+    std::unordered_map<connection_id, lw_shared_ptr<kafka_connection>, boost::hash<connection_id>> _connections;
```

Comment:
Why are kafka_connection objects wrapped into lw_sharde_ptr? object is managed by the connection_manager so I don't understand why we need a smart pointer like that. Why can't we store unique_ptr in the map and return kafka_connection by reference?

Reply:
There is a theoretical risk of the producer going out of scope before every future using the connections finishes. Therefore, we wrap it in a shared_ptr to be 100% safe at a relatively small cost.

Reply:
Read the code again to look for further reasons, we also use this to allow closing connections in the background and then erasing them with finally
`return conn_ptr->close().finally([conn_ptr]{});`

Reply:
I believe the should be a way to do it without shared_ptr and certainly, unique_ptr would be enough. My suspicion is that even unique_ptr is not needed.